### ISSUE TITLE
[RFC] TorchRPC as Single-Controller Backend

### ISSUE THREAD

**Issue by zTonyZhao** [volcengine/verl#1221]
# Motivation

verl was designed in layered architecture, using WorkerGroups with Ray backend as the single-controller of the RL workflow. Ray is well known for its easy expansion and pythonic interface. However, there exist users that are not familiar with Ray, hoping for an alternative backend for verl. Thus, an implementation of WorkerGroup using TorchRPC from PyTorch offers this community a second choice.

# Proposed Add-Ons

New low-level modules including `TorchRPCResourcePool`, `TorchRPCClassWithInitArgs` and `TorchRPCWorkerGroup` will be implemented using TorchRPC in replace of Ray.

We try our best to keep APIs aligned to those in Ray-backend. Certain features like detached workers and named workers will not be implemented due to limitations of TorchRPC.

# Demo

Here follows an equivalent test of [tests/ray/test_worker_group_basics.py](https://github.com/volcengine/verl/tree/main/tests/ray/test_worker_group_basics.py) using torchrpc backend.

No modifications other than module importing are needed, with the `@register` decorator and the dispatch-collect method stay the same.

```python
import os
import torch
import torch.distributed.rpc as rpc

from verl.single_controller.base.decorator import Dispatch, Execute, collect_all_to_all, register
from verl.single_controller.base.worker import Worker
from verl.single_controller.torchrpc import TorchRPCResourcePool, TorchRPCClassWithInitArgs, TorchRPCWorkerGroup, rref_to_here

def two_to_all_dispatch_fn(worker_group, *args, **kwargs):
    """
    Assume the input is a list of 2. Duplicate the input interleaved and pass to each worker.
    """
    for arg in args:
        assert len(arg) == 2
        for i in range(worker_group.world_size - 2):
            arg.append(arg[i % 2])
    for k, v in kwargs.items():
        assert len(v) == 2
        for i in range(worker_group.world_size - 2):
            v.append(v[i % 2])
    return args, kwargs

class TestActor(Worker):
    def __init__(self, x) -> None:
        super().__init__()
        self._x = x

    def foo(self, y):
        return self._x + y

    @register(dispatch_mode=Dispatch.ALL_TO_ALL, execute_mode=Execute.RANK_ZERO)
    def foo_rank_zero(self, x, y):
        return self._x + y + x

    @register(Dispatch.ONE_TO_ALL, blocking=False)
    def foo_one_to_all(self, x, y):
        return self._x + y + x

    @register(Dispatch.ALL_TO_ALL, blocking=False)
    def foo_all_to_all(self, x, y):
        return self._x + y + x

    @register(dispatch_mode={"dispatch_fn": two_to_all_dispatch_fn, "collect_fn": collect_all_to_all})
    def foo_custom(self, x, y):
        return self._x + y + x

def add_one(data):
    data = data.to("cuda")
    data += 1
    data = data.to("cpu")
    return data

def test_basics():
    # MASTER_ADDR, MASTER_PORT, TORCHRPC_RANK, TORCHRPC_WORLD_SIZE must be set first
    # Then run this code on every node
    rank = int(os.environ.get('TORCHRPC_RANK'))
    world_size = int(os.environ.get('TORCHRPC_WORLD_SIZE'))
    rpc.init_rpc(f"worker{rank}", rank=rank, world_size=world_size)
    if rank == 0:
        resource_pool = TorchRPCResourcePool([2,2], use_gpu=True)
        class_with_args = TorchRPCClassWithInitArgs(cls=TestActor, x=2)
        worker_group = TorchRPCWorkerGroup(
            resource_pool=resource_pool, cls_with_init=class_with_args, name_prefix="worker_group_basic"
        )

        output = worker_group.execute_all_sync("foo", y=3)
        assert output == [5, 5, 5, 5]

        # this is a list of object reference. It won't block.
        output_ref = worker_group.execute_all_async("foo", y=4)
        print(output_ref)

        assert rref_to_here(output_ref) == [6, 6, 6, 6]

        output_ref = worker_group.foo_one_to_all(x=1, y=2)
        assert rref_to_here(output_ref) == [5, 5, 5, 5]

        output_ref = worker_group.foo_all_to_all(x=[1, 2, 3, 4], y=[5, 6, 7, 8])
        assert rref_to_here(output_ref) == [8, 10, 12, 14]

        output = worker_group.execute_func_rank_zero(add_one, torch.ones(2, 2))
        torch.testing.assert_close(output, torch.ones(2, 2) + 1)

        output_ref = worker_group.foo_custom(x=[1, 2], y=[5, 6])
        assert output_ref == [8, 10, 8, 10]

        output_ref = worker_group.foo_rank_zero(x=1, y=2)
        assert output_ref == 5
    rpc.shutdown()

if __name__ == "__main__":
    test_basics()
```

# TODO

- [ ] pass equivalent tests in [tests/ray](https://github.com/volcengine/verl/tree/main/tests/ray)
  - [ ] `test_worker_group_basics.py`
  - [ ] `test_worker_group_torch.py`
  - [ ] `test_colocated_workers.py`
  - [ ] `test_data_transfer.py`
  - [ ] `test_driverfunc_to_worker.py`
  - [ ] `test_high_level_scheduling_api.py`
- [ ] pass equivalent test in [tests/e2e/arithmetic_sequence](https://github.com/volcengine/verl/tree/main/tests/e2e/arithmetic_sequence)
  - [ ] Implement `TorchRPCPPOTrainer` parallel to `RayPPOTrainer`

# CC
@eric-haibin-lin @zw0610

**Comment by wuxibin89** [comment]
@zTonyZhao Hi, thanks for your proposal. I'm doubt that TorchRPC has enough flexibility to build a heterogeneous and dynamic RLHF system as we are evolving towards multi-turn, agentic RL.
- heterogeneous: TorchRPC workers are essentially static, homogeneous, and running in SPMD fashion. `rpc.init_rpc` still have rank/world_size which means that they must be launched together and `init_process_group` in advance. But in RLHF system, beside training/inference workers, we have various other workers such as reward server, sandbox server, which has no rank/world_size concept.
- dynamic: as describe above, TorchRPC workers are static, it's hard to dynamically launch and tear down new workers, which prevents parallel compute_score, elastic rollouts.

Besides, TorchRPC lacks of named actor, asyncio actor support, which are very import to build the online LLM serving system.

**Comment by zw0610** [comment]
@wuxibin89 

Let's look forward to how TonyZhao gonna addressed issues like heterogeneous workers.

But asyncio? That's a complete disaster for verl users. I'll try my best to remove these asyncio call from trainer loops.

**Comment by casper-hansen** [comment]
Hi @zw0610. I tested the asyncio implementation by @wuxibin89. I think it's nice and it works great for my use-case of tool-calling. Why are you against working in asyncio? How else would you propose asynchronous rollouts? To give you some context, the previous implementation in veRL was a big disaster because you had to synchronize all samples when doing multi-round generations, causing a lot of issues and unbalanced usage of GPUs.

**Comment by zw0610** [comment]
@casper-hansen 

It is one thing to have asyncio in PPOTrainer for multi-round cases, which I'm not favor. And it is another to to assume it's a must when against introducing another backend for verl. So far as I know, verl serves as an RLHF framework instead of a Ray extension. If you want to discuss whether we should have asyncio in PPOTrainer, I would strongly suggest you start a new issue and leave this one to TorchRPC backend RFC.

P.S. What I said is 'that's a complete disaster for verl users to assume they are writing the rl steps in an asycnio coroutine' instead of simply 'against working in asyncio'.

**Comment by Kirrito-k423** [comment]
### Scheduled issue cleanup / 定时 issue 清理

中文：该 issue 命中定时清理规则：最近一次非当前运行账号的公开回复或引用已经超过 6 个月。因此按定时清理标准先关闭。若作者或社区成员仍能提供新的复现、证据或需求背景，欢迎继续评论请求重开，或创建带有最新信息的新 issue。

证据：
- Issue: https://github.com/verl-project/verl/issues/1221
- 最近一次非当前运行账号的公开回复或公开引用是 `2025-05-15T03:23:55Z` UTC（cross-referenced by @eric-haibin-lin），截至 2026-07-05 已超过 6 个月。
- 已检查 comments 来源：https://api.github.com/repos/verl-project/verl/issues/1221/comments
- 已检查 timeline 来源：https://api.github.com/repos/verl-project/verl/issues/1221/timeline

English: This issue matches the scheduled cleanup rule: the latest public reply or reference by an account other than the current runner is older than 6 months. Therefore, this is being closed under the scheduled cleanup policy. If the author or community members can still provide a fresh reproduction, evidence, or requirement context, please comment to request reopening or open a new issue with updated information.

Evidence:
- Issue: https://github.com/verl-project/verl/issues/1221
- The latest public reply or public reference from an account other than the current runner was `2025-05-15T03:23:55Z` UTC (cross-referenced by @eric-haibin-lin), which is more than 6 months before 2026-07-05.
- Checked comments source: https://api.github.com/repos/verl-project/verl/issues/1221/comments
- Checked timeline source: https://api.github.com/repos/verl-project/verl/issues/1221/timeline

This response was generated by a scheduled Codex GPT5.5 issue processing task.
